util/src/main/java/io/kubernetes/client/util/Cache.java

```diff
@@ -0,0 +1,157 @@
+package io.kubernetes.client.util;
+
+import com.google.gson.reflect.TypeToken;
+import com.squareup.okhttp.Call;
+import com.squareup.okhttp.ResponseBody;
+import io.kubernetes.client.ApiClient;
+import io.kubernetes.client.ApiException;
+import io.kubernetes.client.models.V1ObjectMeta;
+import java.io.Closeable;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.function.Function;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Cache<ApiType, ApiListType> implements Closeable {
+  private Function<Boolean, Call> callGenerator;
+  private ApiClient client;
+  private Watch<ApiType> watch;
+  private ApiListType list;
```

Comment:
we still need to index the list, or a namespaced list from a all-namespace cache will be costy

Reply:
yep.  There are lots of optimizations we can/should do.